db/view/view.cc

```diff
@@ -3780,6 +3798,107 @@ bool view_building_worker::is_shard_free(shard_id shard) {
+            std::vector<canonical_mutation> cmuts;
+
+            for (auto& task: tasks) {
+                // Generate new id for the task
+                task.id = utils::UUID_gen::get_time_UUID();
+                auto mut = co_await _group0_client.sys_ks().make_view_building_task_mutation(guard.write_timestamp(), task);
+                cmuts.emplace_back(std::move(mut));
+            }
+
+            auto cmd = _group0_client.prepare_command(service::write_mutations{std::move(cmuts)}, guard, "create view building tasks");
+            co_await _group0_client.add_entry(std::move(cmd), std::move(guard), _as);
+            break;
+        } catch (service::group0_concurrent_modification&) {
+            vbw_logger.warn("Got group0_concurrent_modification error in save_view_building_tasks(). Retries left: {}", retries_left);
+            if (retries_left == 0) {
+                std::rethrow_exception(std::current_exception());
+            }
+        }
+    }
+}
+
+static locator::tablet_id get_sstable_tablet_id(const locator::tablet_map& tablet_map, const sstables::sstable& sst) {
+    auto last_token = sst.get_last_decorated_key().token();
+    auto tablet_id = tablet_map.get_tablet_id(last_token);
+
+#ifdef SEASTAR_DEBUG
+    // Single sstable from tablet-table should contain data for only one tablet
+    auto first_token = sst.get_first_decorated_key().token();
+    auto first_token_tablet_id = tablet_map.get_tablet_id(first_token);
+    SCYLLA_ASSERT(tablet_id == first_token_tablet_id);
+#endif
+
+    return tablet_id;
+}
+
+static shard_id get_sstable_shard_id(const sstables::sstable& sst) {
+    auto shards = sst.get_shards_for_this_sstable();
+#ifdef SEASTAR_DEBUG
+    // Sstable from tablet-table should belong to only one shard
+    SCYLLA_ASSERT(shards.size() == 1);
+#endif
+    return shards[0];
+}
+
+static bool staging_task_exists(const service::view_building::building_tasks& tasks, table_id table_id, const locator::tablet_replica& replica, locator::tablet_id tid) {
+    if (!tasks.contains(table_id) || !tasks.at(table_id).contains(replica)) {
+        return false;
+    }
+    auto& replica_tasks = tasks.at(table_id).at(replica);
+    return std::ranges::any_of(replica_tasks.staging_tasks, [&tid] (auto& t) {
+        return t.second.tid == tid;
+    });
+}
+
+// Because view building state lives only on shard0, the method needs to take copy of building tasks
+// to detemine whether a task for particular staging sstable exists or not.
+future<> view_building_worker::discover_staging_sstables(service::view_building::building_tasks building_tasks) {
+    std::vector<service::view_building::view_building_task> tasks_to_create;    
+    auto my_host_id = _db.get_token_metadata().get_topology().my_host_id();
+
+    _db.get_tables_metadata().for_each_table([&] (table_id table_id, lw_shared_ptr<replica::table> table) {
+        if (!table->uses_tablets()) {
+            return;
```

Comment:
Is group0's view build coordinator mutually exclusive with vnodes? Otherwise, we might want to support it too. Is it enabled via a feature flag? IIRC, we today support a cluster with both vnodes and tablets, although it's not what we want in the long run, but it can happen. That being said, can we have both today's local view building working concurrently with the group0 one, to support both vnode and tablet?